Mathlib/Computability/QueryComplexity/Defs.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2025 Tomaz Mascarenhas. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Geoffrey Irving, Tomaz Mascarenhas
+-/
+import Mathlib.Algebra.Group.Pi.Basic
+import Mathlib.Data.Set.Basic
+
+/-!
+# Definition of a model of computation based on oracles.
+
+This module defines an abstraction of computation with oracles, enabling the analysis of upper
+bounds on the query complexity of algorithms. It also provides a Monad instance for it.
+
+Note that this is the deterministic version. In the future, the stochastic version will
+also be ported, which enables the possibility of proving lower bounds on the query complexity
+of algorithms. In order to facilitate this future integration, the design of the types
+in this module is based on the original stochastic version. Source:
+  https://github.com/girving/debate/blob/862fdb1cf55df0d541b802bdb1e672d724df6398/Comp/Oracle.lean
+
+## Main Definitions
+
+* Comp (ι : Type) {I : Type} (s : Set I) (α : Type) : Type
+* Comp.run : Comp ι s α → (I → Oracle ι) → α × (I → ℕ)
+-/
+
+open Classical
+open Set
+noncomputable section
+
+variable {ι I : Type}
+variable {s t : Set I}
+variable {α β γ : Type}
+
+namespace QueryComplexity
+
+/-- A deterministic oracle is a map from `α` to `Bool` -/
+def Oracle (α : Type) := α → Bool
+
+/-- A deterministic computation that make decisions by querying oracles. A computation is either a
+pure value or the identifier of an oracle (`o`) drawn from a predefined set (`s`), a value to
+be queried by the oracle (`i`) and two other computations, to be run depending on the answer
+of the oracle. -/
+inductive Comp (ι : Type) {I : Type} (s : Set I) (α : Type) : Type where
+  | pure' : α → Comp ι s α
+  | query' : (o : I) → o ∈ s → (i : ι) → (Bool → Comp ι s α) → Comp ι s α
+
+namespace Comp
+
+/-- The standard bind operation for `Comp` -/
+def bind' (f : Comp ι s α) (g : α → Comp ι s β) : Comp ι s β := match f with
+  | .pure' x => g x
+  | .query' o m y f => .query' o m y fun b => (f b).bind' g
+
+/-- `Comp` is a monad -/
+instance : Monad (Comp ι s) where
+  pure := Comp.pure'
+  bind := Comp.bind'
+
+/-- Produce a `Comp` given the identifier of an oracle and a value to be queried. The `Comp`
+just returns `true` or `false` according to the answer of the oracle. -/
+def query (o : I) (y : ι) : Comp ι {o} Bool :=
+  Comp.query' o (mem_singleton _) y pure
+
+/-- Execute `f` with the oracles `os`. Returns the final value and the number of queries to
+each one of the oracles. -/
+def run (f : Comp ι s α) (os : I → Oracle ι) : α × (I → ℕ) := match f with
+  | .pure' x => (x, fun _ => 0)
+  | .query' i _ y f =>
+    let x := os i y
+    let (z,c) := (f x).run os
+    (z, c + fun j => if j = i then 1 else 0)
```

Comment:
```suggestion
    (z, c + Pi.single i 1)
```

Reply:
Though using `DFinsupp` here would be slightly nicer because then you get a computable `Repr` for free

Reply:
I think `DFinsupp` would be less ergonomic in practice, since it's harder to write down functions then.

Unfortunately I think a meta-lesson here is that the needs of the particular `debate` formalisation diverge from the maximally general version of `Comp`. E.g., I started doing the dependent output stuff, and it'll be fine mostly, but then all my applications will be `Bool` for fundamental reasons: the theorems depend on the number of bits in the representation of the output type, and it's cleaner if the number of bits is one.

On the margin that argues against trying to upstream it at this point, especially if the only standard (non-debate) result is sorting.

Reply:
In light of https://github.com/leanprover-community/mathlib4/pull/20924#discussion_r1929625674, are you in favor of the generality again?

Reply:
(I also think it's fine to not use DFinsupp for now)

Reply:
Yep, output generality seems very good: it turned out to be easy to migrate to.

Reply:
@tomaz1502, note the code suggestion above is independent of the radio of the discussion.

Reply:
I took the liberty of applying this, and also adding `DecidableEq` because it seems convenient for computations to actually be computable.